src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/LeoAuthProviderHelper.scala

```diff
@@ -32,48 +35,51 @@ object LeoAuthProviderHelper {
   }
 }
 
-class LeoAuthProviderHelper(wrappedAuthProvider: LeoAuthProvider, authConfig: Config, serviceAccountProvider: ServiceAccountProvider) extends LeoAuthProvider(authConfig, serviceAccountProvider) with LazyLogging {
+class LeoAuthProviderHelper(val wrappedProvider: LeoAuthProvider, authConfig: Config, serviceAccountProvider: ServiceAccountProvider)(implicit val system: ActorSystem)
+  extends LeoAuthProvider(authConfig, serviceAccountProvider) with SamProviderHelper[LeoAuthProvider] {
 
-  private def safeCall[T](future: => Future[T])(implicit executionContext: ExecutionContext): Future[T] = {
-    val exceptionHandler: PartialFunction[Throwable, Future[Nothing]] = {
-      case e: LeoException => Future.failed(e)
-      case NonFatal(e) =>
-        val wrappedClassName = wrappedAuthProvider.getClass.getSimpleName
-        logger.error(s"Auth provider $wrappedClassName throw an exception", e)
-        Future.failed(AuthProviderException(wrappedClassName))
-    }
-
-    // recover from failed futures AND catch thrown exceptions
-    try { future.recoverWith(exceptionHandler) } catch exceptionHandler
-  }
+  // Cache notebook auth results from Sam as this is called very often by the proxy and the "list clusters" endpoint.
+  // Project-level auth is not called as frequently so it's not as important to cache it.
+  private val notebookAuthCache = CacheBuilder.newBuilder()
+    .expireAfterAccess(5, TimeUnit.MINUTES)
```

Comment:
`expireAfterAccess` or `expireAfterWrite`? Hmmm

Reply:
After write, otherwise repeatedly trying to access a thing will maintain your access to it even if someone removes you.

Reply:
Makes sense